Display notice when users record or transcribe meeting

cheukchuen opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
Display notification banner when users record and/or transcribe meeting

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

cheukchuen committed:
Display notice when users record or transcribe meeting

waynemo commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -521,10 +522,46 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
     @objc func recordingActiveChangeUpdated(_ notification: Notification) {
-        guard let isRecordingActive = callingContext?.isRecordingActive else {
+        guard let isRecordingActive = callingContext?.isRecordingActive,
+              let isTranscriptionActive = callingContext?.isTranscriptionActive else {
             return
         }
-        let notificationText = isRecordingActive ? meetingRecordingActiveText : meetingRecordingStopText
+
+        let notificationText: NSAttributedString = {
+            switch (isRecordingActive, isTranscriptionActive) {
+            case (true, false):
+                return meetingRecordingActiveText
+            case (true, true):
+                return meetingRecordingAndTranscriptionActiveText
+            case (false, true):
+                return meetingRecordingStopTranscriptionActiveText
+            default:
+                return meetingRecordingStopText
+            }
+        }()
+
+        messageBannerView.showBannerMessage(notificationText)
```
We can create a private func and reuse this logic for both.

AmyL219 replied:
Not sure if one logic function can be applied to the events.
Since the text is different based on triggering event. For example, if isRecordingActive = true, isTranscriptionActive = false. If it is in the Recording listener, we would display "Recording start", if it's trigger from Transcription change, it would display "Transcription stops but recording is continuing"

waynemo replied:
We always update the message based on the values of both `isRecordingActive` and `isTranscriptionActive`, and at this point those values are already updated in `CallingContext`. So we should be able to reuse the same logic.

cheukchuen replied:
Updated with a helper function. Also changed the base branch to target `feature/teams_interop`

cheukchuen committed:
Reuse helper function when toggling recording and transcription

cheukchuen committed:
Consistent input parameters when toggling notification

AmyL219 commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -520,14 +521,44 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
-    @objc func recordingActiveChangeUpdated(_ notification: Notification) {
-        guard let isRecordingActive = callingContext?.isRecordingActive else {
+    private func toggleMeetingCaptureNotification(recording: NSAttributedString, transcribing: NSAttributedString, bothActive: NSAttributedString, stopped: NSAttributedString) {
+        guard let isRecordingActive = callingContext?.isRecordingActive,
```
I think we could combine the word logic into the toggleNotification function
You could have a look on Android side implement method (use an enum as input) or implement a simple `state machine`: https://albertodebortoli.com/2018/12/16/the-easiest-state-machine-in-swift/

cheukchuen replied:
Could you please clarify the Android side implementation method using enums? I'm seeing basic implementation using [if-else logic](https://github.com/Azure-Samples/communication-services-android-calling-hero/blob/2d529525085f759fa02f16efa7431f742d5a0170/AzureCalling/app/src/main/java/com/azure/samples/communication/calling/activities/CallActivity.java#L378-L404).

Also wondering if we are to use state machines, could we just utilize the [Stateful](https://github.com/albertodebortoli/Stateful) package by including `pod 'Stateful'` in the podfile?

cheukchuen replied:
I've changed the `toggleMeetingCaptureNotification` method to accept an enum. That way if there are more types added in the future, the switch statement grows instead of the parameter list of a function.

Personally I would prefer an actual state machine implementation, but for now that might be an overkill.

cheukchuen committed:
Refactor toggleMeetingCaptureNotification to accept a single enum param

cheukchuen committed:
Rename to messageBannerStackView in CallViewController

cheukchuen committed:
Merge branch 'feature/teams_interop' into csiow/recording-transcription-notice

AmyL219 reviewed (dismissed):
Looks good to me

waynemo commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -520,12 +526,40 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
-        let notificationText = isRecordingActive ? meetingRecordingActiveText : meetingRecordingStopText
-        messageBannerView.showBannerMessage(notificationText)
+
+        let notificationText: NSAttributedString = {
+            switch (mode, isRecordingActive, isTranscriptionActive) {
+            case (_, true, true):
+                return meetingRecordingAndTranscriptionActiveText
+            case (.recording, true, false):
+                return meetingRecordingActiveText
+            case (.recording, false, true):
+                return meetingRecordingStopTranscriptionActiveText
+            case (.recording, false, false):
+                return meetingRecordingStopText
+            case (.transcription, true, false):
+                return meetingRecordingActiveTranscriptionStopText
+            case (.transcription, false, true):
+                return meetingTranscriptionActiveText
+            case (.transcription, false, false):
+                return meetingTranscriptionStopText
+            }
+        }()
```
Creating this closure every time just to return the text does not feel right. I think we can just assign the `notificationText` variable to simplify it.

waynemo commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -35,7 +40,7 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
     @IBOutlet weak var toggleMuteButton: UIButton!
     @IBOutlet weak var infoHeaderView: InfoHeaderView!
-    @IBOutlet weak var messageBannerView: MessageBannerStackView!
+    @IBOutlet weak var messageBannerStackView: MessageBannerStackView!
```
Since we are using the name NoticeBannerType, I think we can rename this to `noticeBannerStackView` and `NoticeBannerStackView` to be consistent.

waynemo commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -520,12 +526,40 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
 
-    @objc func recordingActiveChangeUpdated(_ notification: Notification) {
-        guard let isRecordingActive = callingContext?.isRecordingActive else {
+    private func toggleMeetingCaptureNotification(mode: NoticeBannerType) {
```
To align with the naming, we can call this `showNoticeBanner`.

cheukchuen committed:
Rename to NoticeBannerStackView

cheukchuen committed:
Update showNoticeBanner

- Simple string assignment for notificationText instead of using closure to return value

waynemo reviewed (approved):
Looks good to me.

Let's also make sure to squash merge to feature/teams_interop and clean up some of the intermediate messages.
